### DISCUSSION TITLE
static and shared internal dependencies

### DISCUSSION THREAD

**Discussion by bruchar1** [mesonbuild/meson#12007] [Q&A]
My project uses a lot of internal libraries, and I use `declare_dependency` to express the transitive dependencies between those libraries. Normally, all the dependencies are shared libraries. However, my project have plugin libraries, that depend on the same common libraries, but I want to link statically, because I may update a plugin in an installation without updating the executable and all its shared libraries.

How can I express the fact that a dependency can be either a static library depending on other static libraries, or a shared library, depending on shared libraries?

What I'd like could be, for example, to have a dependency on a `both_libraries` target, and to select the right version (static or shared) or the lib and of all its dependencies. Should we have a `declare_both_dependency` or something like that to achieve this? Or is there a simpler way? Because, actually, I need to duplicate a lot of code in my meson.build files...

**Comment by xclaesse** [comment]
This is a common issue that I think we need to find a solution. Typically a library project does:
```meson
l = library()
dep = declare_dependency(link_with: l)
```

Now if building with `--default-library=both`, or if using `both_libraries()` directly, using `dep` will always use the shared library. It would be nice to add `dep.as_static()` that recursively crawl all link_with and if it finds any BothLibraries, replace it with its StaticLibrary part. This would be similar to `as_link_whole()` we already have: https://mesonbuild.com/Reference-manual_returned_dep.html#depas_link_whole.

**Reply by xclaesse** [reply]
I've been wanting something like this to allow building GStreamer shared plugins and gst-full (a shared library that static link all libs and plugins) at the same time.

**Reply by bruchar1** [reply]
#12632 should provide a solution for this.
